File: internal/transport/lambda.go
package transport

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/aws/aws-lambda-go/events"

	"mcpserver"
)

// LambdaAdapter wraps an MCP Server for AWS Lambda API Gateway V2.
type LambdaAdapter struct {
	server *mcpserver.Server
}

// NewLambdaAdapter creates a Lambda adapter for the given MCP server.
func NewLambdaAdapter(server *mcpserver.Server) *LambdaAdapter {
	return &LambdaAdapter{server: server}
}

// Handle satisfies the aws-lambda-go handler signature for API Gateway V2 HTTP APIs.
func (a *LambdaAdapter) Handle(ctx context.Context, request events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
	method := request.RequestContext.HTTP.Method
	path := request.RawPath

	switch {
	case strings.EqualFold(method, http.MethodGet) && path == "/healthz":
		return events.APIGatewayV2HTTPResponse{
			StatusCode: http.StatusOK,
			Headers:    map[string]string{"content-type": "application/json"},
			Body:       `{"status":"ok"}`,
		}, nil

	case strings.EqualFold(method, http.MethodPost) && path == "/mcp":
		return a.handleJSONRPC(ctx, request)

	default:
		return events.APIGatewayV2HTTPResponse{
			StatusCode: http.StatusNotFound,
			Headers:    map[string]string{"content-type": "application/json"},
			Body:       `{"error":"route_not_found"}`,
		}, nil
	}
}

func (a *LambdaAdapter) handleJSONRPC(ctx context.Context, request events.APIGatewayV2HTTPRequest) (events.APIGatewayV2HTTPResponse, error) {
	body := request.Body
	if request.IsBase64Encoded {
		decoded, err := base64.StdEncoding.DecodeString(body)
		if err != nil {
			return lambdaError(http.StatusBadRequest, fmt.Errorf("base64 decode: %w", err)), nil
		}
		body = string(decoded)
	}

	var rpcReq mcpserver.JSONRPCRequest
	if err := json.Unmarshal([]byte(body), &rpcReq); err != nil {
		resp := mcpserver.NewErrorResponse(nil, mcpserver.ErrCodeParse, "invalid JSON: "+err.Error())
		return lambdaJSON(resp, http.StatusBadRequest), nil
	}

	// Notifications produce no response body.
	if strings.HasPrefix(rpcReq.Method, "notifications/") {
		return events.APIGatewayV2HTTPResponse{StatusCode: http.StatusAccepted}, nil
	}

	resp := a.server.Handle(ctx, rpcReq)

	if resp.IsNotification() {
		return events.APIGatewayV2HTTPResponse{StatusCode: http.StatusAccepted}, nil
	}

	return lambdaJSON(resp, http.StatusOK), nil
}

func lambdaJSON(payload any, status int) events.APIGatewayV2HTTPResponse {
	buf, err := json.Marshal(payload)
	if err != nil {
		return events.APIGatewayV2HTTPResponse{
			StatusCode: http.StatusInternalServerError,
			Headers:    map[string]string{"content-type": "application/json"},
			Body:       fmt.Sprintf(`{"error":"%s"}`, err),
		}
	}
	return events.APIGatewayV2HTTPResponse{
		StatusCode: status,
		Headers:    map[string]string{"content-type": "application/json"},
		Body:       string(buf),
	}
}

func lambdaError(status int, err error) events.APIGatewayV2HTTPResponse {
	return events.APIGatewayV2HTTPResponse{
		StatusCode: status,
		Headers:    map[string]string{"content-type": "application/json"},
		Body:       fmt.Sprintf(`{"error":"%s"}`, err),
	}
}

File: rust/crates/marketplace/src/lambda.rs
use std::sync::Arc;

use lambda_http::{Request, Response, Body, Error, service_fn};
use mcpserver::{JsonRpcRequest, Server, new_error_response};

/// Run the Lambda handler loop.
pub async fn run(srv: Server) {
    let srv = Arc::new(srv);
    let func = service_fn(move |event: Request| {
        let srv = srv.clone();
        async move { handle(event, &srv).await }
    });
    lambda_http::run(func).await.unwrap();
}

async fn handle(event: Request, srv: &Server) -> Result<Response<Body>, Error> {
    let method = event.method().as_str().to_uppercase();
    let path = event.uri().path();

    match (method.as_str(), path) {
        ("GET", "/healthz") => {
            Ok(Response::builder()
                .status(200)
                .header("content-type", "application/json")
                .body(Body::Text(r#"{"status":"ok"}"#.into()))
                .unwrap())
        }
        ("POST", "/mcp") => handle_jsonrpc(event, srv).await,
        _ => {
            Ok(Response::builder()
                .status(404)
                .header("content-type", "application/json")
                .body(Body::Text(r#"{"error":"route_not_found"}"#.into()))
                .unwrap())
        }
    }
}

async fn handle_jsonrpc(event: Request, srv: &Server) -> Result<Response<Body>, Error> {
    let body = match event.body() {
        Body::Text(s) => s.clone(),
        Body::Binary(b) => String::from_utf8_lossy(b).into_owned(),
        Body::Empty => String::new(),
    };

    let rpc_req: JsonRpcRequest = match serde_json::from_str(&body) {
        Ok(r) => r,
        Err(e) => {
            let resp = new_error_response(
                None,
                -32700, // parse error
                format!("invalid JSON: {}", e),
            );
            let json = serde_json::to_string(&resp).unwrap();
            return Ok(Response::builder()
                .status(400)
                .header("content-type", "application/json")
                .body(Body::Text(json))
                .unwrap());
        }
    };

    // Notifications produce no response body.
    if rpc_req.method.starts_with("notifications/") {
        return Ok(Response::builder()
            .status(202)
            .body(Body::Empty)
            .unwrap());
    }

    let resp = srv.handle(rpc_req).await;

    // Check if the response is a notification sentinel.
    if resp.id.is_none() && resp.result.is_none() && resp.error.is_none() {
        return Ok(Response::builder()
            .status(202)
            .body(Body::Empty)
            .unwrap());
    }

    let json = serde_json::to_string(&resp).unwrap();
    Ok(Response::builder()
        .status(200)
        .header("content-type", "application/json")
        .body(Body::Text(json))
        .unwrap())
}
